openwisp_firmware_upgrader/base/models.py

```diff
@@ -612,6 +613,11 @@ def log_line(self, line, save=True):
         if save:
             self.save()
 
+        publisher = UpgradeProgressPublisher(self.pk)
```

Comment:
`self.pk` can be `None` if the model instance hasn’t been saved yet (i.e., it’s a new object). Also, I believe `UpgradeProgressPublisher()` expects a non-null `operation_id`. :thinking: 
```suggestion
     if self.pk:
            publisher = UpgradeProgressPublisher(self.pk)
            publisher.publish_progress(
                {"type": "log", "content": line, "status": self.status}
            )
```

Reply:
Right!
Will add this check
Thanks